src/Closure.cpp

```diff
@@ -106,5 +109,79 @@ void Closure::visit(const Atomic *op) {
     op->body.accept(this);
 }
 
+Expr Closure::pack_into_struct() const {
+    std::vector<Expr> elements;
+    std::vector<Type> types;
+
+    std::ostringstream name;
+
+    for (const auto &b : buffers) {
+        Expr ptr_var = Variable::make(type_of<void *>(), b.first);
+        elements.emplace_back(ptr_var);
+    }
+    for (const auto &v : vars) {
+        Expr var = Variable::make(v.second, v.first);
+        elements.emplace_back(var);
+    }
+
+    // Sort by decreasing size, to guarantee the struct is densely packed in
+    // memory. We don't actually rely on this, it's just nice to have.
+    std::stable_sort(elements.begin(), elements.end(),
+                     [&](const Expr &a, const Expr &b) {
+                         return a.type().bytes() > b.type().bytes();
+                     });
+
+    for (Expr e : elements) {
+        types.push_back(e.type());
+        name << e.type() << "_";
+    }
+
+    Expr result = Call::make(Handle(),
+                             Call::make_struct, elements, Call::Intrinsic);
+    return result;
+}
+
+Stmt Closure::unpack_from_struct(const Expr &e, const Stmt &s) const {
+    // Use the struct-packing code just to make sure the order of elements is
+    // the same.
+    Expr packed = pack_into_struct();
+
+    // Make a prototype of the packed struct
+    class ReplaceCallArgsWithZero : public IRMutator {
+    public:
+        using IRMutator::mutate;
+        Expr mutate(const Expr &e) override {
+            if (!e.as<Call>()) {
+                return make_zero(e.type());
+            } else {
+                return IRMutator::mutate(e);
+            }
+        }
+    } replacer;
+    string prototype_name = unique_name("closure_prototype");
```

Comment:
random thought: any reason not to use single-char shortnames here (e.g. `unique_name('c')`? this is here presumably to improve readability of our IR, but if that's the goal, we should try to also include some name unique to the specific final name of the Closure's loop/lambda/etc; otherwise something like `c$3` is arguably just as good.

Reply:
It's a confusing thing in the IR. I think it deserves a long name to say why there's an apparently unused struct.